Richer font highlighting

theothornhill opened the pull request:
Font locking is inspired by vs code and how it fontifies quite uniformly.
However, we still adhere to normal CC Mode conventions, and therefore deviate in
quite a few spots.

theothornhill commented:
I only made tests pass, since I'm thinking of rewriting the whole test suite. I hope that is OK for you :)

#177, #180, #181 is fixed after this commit

josteink commented on csharp-mode.el:
```diff
@@ -146,7 +146,7 @@
 (c-lang-defconst c-other-kwds
   csharp '("select" "from" "where" "join" "in" "on" "equals" "into"
-           "orderby" "descending" "group" "nameof"))
+           "orderby" "descending" "group" "nameof" "when"))
```
If we are to support all LinQ keywords, I believe we are missing "let" in this list.

https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/let-clause

theothornhill replied:
Added in 82bcec9

josteink replied:
Yeah. That's much better, and we seemingly missed quite a few more than I noticed :laughing:

josteink commented on csharp-mode.el:
```diff
@@ -285,12 +284,27 @@ casts and declarations are fontified.  Used on level 2 and higher."
+          `(("\\[\\([A-Z][A-Za-z0-9_]*\\)\\][^;]" 1 font-lock-variable-name-face t))
+
+          ;;  Types after 'new'
+          `((,(concat "new\\s *\\([" (c-lang-const c-symbol-chars) "]+\\)")
```
Any reason why you are using `\\([A-Z][A-Za-z0-9_]+\\)` for types elsewhere, but here you are using `\\([" (c-lang-const c-symbol-chars) "]+\\)` ?

Also: Should we define our own constant, for instance `csharp-re-typename` ?

josteink replied:
Speaking of which... I see you are using a variation of these 2 regexps:

* `\\([A-Z][A-Za-z0-9_]+\\)`
* `\\([A-Z][A-Za-z0-9_]*\\)`

The first variant forces a type-name to have to be 2-characters or more, while the second allows single-character type-names, like `T`. I think this is the more correct regexp to use, unless you have specific reasons not to.

And seeing inconsistencies like this in the code, again makes me think defining this as a single constant is a good idea.

theothornhill replied:
Fixed in 91c0f3c - do you agree?

theothornhill replied:
Yeah, I saw that when going over it again. Thanks for pointing it out. Should be clearer now. These inconsistencies come from hacking on small variations over and over again 😄 

I think not using the `c-symbol-chars` is the correct way to go, since we have control over that, and it is a simple regex.

josteink replied:
Looks good to me now :+1:

josteink commented on csharp-mode.el:
```diff
@@ -432,13 +446,12 @@ casts and declarations are fontified.  Used on level 2 and higher."
 	      (concat "\\sw\\|\\s \\|[=\n\r*.:]\\|"
 		      "\"[^\"]*\"\\|'[^']*'")
-	      "\\)*>")
+	      "\\)*/?>")
```
Assuming this is to support the syntax for nullable types? If so, might be worth a tiny comment to document just that.

theothornhill replied:
In fact, it isn't. This is in the codedoc to match an optional `/>` for the self closing tag.

```xml
<Tag>content</Tag>
<Tag />
```

josteink replied:
Oh yeah, when I see it in a greater context (not just the diff) that becomes immediately obvious. My bad.

josteink commented on csharp-mode.el:
```diff
@@ -285,12 +284,27 @@ casts and declarations are fontified.  Used on level 2 and higher."
 
 (defcustom csharp-font-lock-extra-types
-  (list (concat "[" c-upper "]\\sw*[" c-lower "]\\sw"))
+  (list "[A-Z][a-zA-Z0-9_]*")
```
Another place where we might benefit from having a regexp constant for type-names.

theothornhill replied:
Added in 91c0f3c

josteink commented:
I've made a few comments regarding the actual major-mode implementation and how I think that can be made just *slightly* better.

> I only made tests pass, since I'm thinking of rewriting the whole test suite. I hope that is OK for you :)

Absolutely fair. As such (contrary to what I normally do :cowboy_hat_face: ) I haven't put any effort into reviewing those.

theothornhill commented:
Added a couple new commits to address your comments. Also, by 'making them pass', I've corrected failing tests, not glossed over them. What I haven't done yet is adding new tests that never have existed. Just to be clear 😄

josteink commented:
Looks good to me. Let's get this merged!
